bigtable/google/cloud/bigtable/table.py

```diff
@@ -363,6 +375,74 @@ def sample_row_keys(self):
         return response_iterator
 
 
+class _MutateRowsRetryableError(Exception):
+    """A retryable error in Mutate Rows response."""
+    pass
+
+
+class _RetryableMutateRowsWorker(object):
+    # pylint: disable=unsubscriptable-object
+    RETRY_CODES = (
+        StatusCode.DEADLINE_EXCEEDED.value[0],
+        StatusCode.ABORTED.value[0],
+        StatusCode.UNAVAILABLE.value[0],
+    )
+
+    def __init__(self, client, table_name, rows):
+        self.client = client
+        self.table_name = table_name
+        self.rows = rows
+        self.responses_statuses = [
+            None for _ in six.moves.xrange(len(self.rows))]
+
+    def __call__(self):
+        return self._do_mutate_retryable_rows()
+
+    def _is_retryable(self, status):  # pylint: disable=no-self-use
+        return (status is None or
+                status.code in _RetryableMutateRowsWorker.RETRY_CODES)
+
+    def _next_retryable_row_index(self, begin_index):
+        i = begin_index
+        while i < len(self.responses_statuses):
+            status = self.responses_statuses[i]
+            if self._is_retryable(status):
+                return i
+            i += 1
+        return -1
+
+    def _do_mutate_retryable_rows(self):
+        retryable_rows = []
+        for i, status in enumerate(self.responses_statuses):
+            if self._is_retryable(status):
+                retryable_rows.append(self.rows[i])
+
+        if not retryable_rows:
+            # All mutations are either successful or non-retryable now.
+            return self.responses_statuses
+
+        mutate_rows_request = _mutate_rows_request(
+            self.table_name, retryable_rows)
+        responses = self.client._data_stub.MutateRows(
+            mutate_rows_request)
+
+        num_retryable_responses = 0
+        for response in responses:
+            index = 0
+            for entry in response.entries:
+                index = self._next_retryable_row_index(index)
```

Comment:
This assumes that the responses and the entries will be in order (which is currently how the server is implemented: there is a single response and all entries are sorted by index). However, the protocol allows for async streaming responses from tablets that can be out of order. It would be better to use response.entries[].index for correlating the response to original request.

Reply:
Thanks. I made an assumption that wasn't stated. Fixed in the next commit.